Using kts files instead of .gradle

pfmaggi opened the pull request:
## :loudspeaker: Type of change
<!--- Put an `x` in the boxes that apply -->
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring


## :scroll: Description
Moving to Kotlin DSL for the gradle build scripts.

## :bulb: Motivation and Context
Having a single language both for the app and for its build script.

## :green_heart: How did you test it?
`./gradlew check`

## :pencil: Checklist
<!--- Put an `x` in the boxes that apply -->
- [x] I ran `./gradlew spotlessApply` before submitting the PR
- [x] I reviewed submitted code
- [ ] I added tests to verify changes
- [ ] All tests passing


## :crystal_ball: Next steps
Improve KTS design and build speed.

## :camera_flash: Screenshots / GIFs

pfmaggi committed:
Initial incomplete conversion to Gradle Kotlin DSL

pfmaggi committed:
Completed, not working,  conversion to Kotlin DSL

pfmaggi committed:
Corrected syntax for dynamicFeatures

pfmaggi committed:
More conversion for KTS and spotless Update

pfmaggi committed:
spotless also checks *.kts files

pfmaggi committed:
Updated copyright of *.gradle.kts files

pfmaggi committed:
Enable kapt build cache for :app

pfmaggi committed:
Align import syntax of properties to the one used in Groovy

pfmaggi committed:
Downgraded to AGP v3.6-beta05 to avoid issues on the check task.

pfmaggi committed:
Update to handle properties with double quotes.

florina-muntenescu commented:
would you mind re-basing when you get the chance, please

pfmaggi committed:
Update to AGP 3.6.0-rc1

This commit also correct double quotes handling for properties.

pfmaggi committed:
Add the main Glide artifact to kapt classpath

This is required having disabled compile avoidance for kapt in master:
kapt.include.compile.classpath=false

pfmaggi commented:
Rebase on master plus, a commit to add the main Glide artifact to kapt classpath. This is required having disabled compile avoidance for kapt in master:
```
kapt.include.compile.classpath=false
```

pfmaggi committed:
Revert change to read properties WITH double-quotes.

florina-muntenescu commented on core/proguard-rules.pro:
```diff
@@ -1,6 +1,6 @@
 # Add project specific ProGuard rules here.
 # You can control the set of applied configuration files using the
-# proguardFiles setting in build.gradle.
+# proguardFiles setting in build.gradle.kts.kts.kts.
```
too many `kts`

pfmaggi replied:
Fixed.
Thanks

florina-muntenescu commented on gradle.properties:
```diff
@@ -13,20 +13,20 @@ org.gradle.daemon=true
-
-# Use kapt in parallel
-kapt.use.worker.api=true
+## Use kapt incremental
```
is this not needed anymore?

pfmaggi replied:
Incremental annotation processing is enabled by default starting from version 1.3.50.

florina-muntenescu reviewed (commented):
Overall, for someone with no experience with kts, this looks good. 
Maybe someone else, could take a look as well.

pfmaggi committed:
Integrate some ideas from #811

- Split the config object into three separate files based on content
- Cleanup main build.gradle.kts

pfmaggi committed:
Address a couple of comments on the PR

- Incremental annotation processing is now enabled by default
- Too many kts in "build.gradle.kts.kts.kts".

pfmaggi committed:
Adopt the Java Platform plugin to share dependencies versions.

pfmaggi committed:
Manage kapt version sets through Java Platform plugin

pfmaggi committed:
Adds Libs object with libray names and version set for kapt

keyboardsurfer commented on app/build.gradle.kts:
```diff
@@ -0,0 +1,108 @@
+ */
+
+plugins {
+    id("com.android.application")
```
Sort these please

pfmaggi replied:
The actual order is relevant.
The other option is to use only `id()` for the plugins... but I don't think it's worthwhile.

keyboardsurfer commented on buildSrc/src/main/java/Versions.kt:
```diff
@@ -0,0 +1,25 @@
+ * limitations under the License.
+ */
+
+@Suppress("MayBeConstant") // Improve perf when changing values
```
How does performance get improved here?

pfmaggi replied:
Good point, I think I got that updating this over time... removing it.

keyboardsurfer commented on depconstraints/build.gradle.kts:
```diff
@@ -0,0 +1,121 @@
+
+dependencies {
+    constraints {
+        api(Libs.AX_ANNOTATION + androidx)
```
I prefer Kotlin's string templates over concatenation.

pfmaggi replied:
Fine for me.
Converted

keyboardsurfer commented on depconstraints/build.gradle.kts:
```diff
@@ -0,0 +1,121 @@
+        api(Libs.TEST_UIAUTOMATOR + ui_automator)
+
+        // Adding this to bring "google_play_services_version" into the test project
+        // without this, it fails on AGP 3.6.x.
```
3.6.x+ ?

pfmaggi replied:
This was something required when we switched to AS 3.6, seems fine now, removing.

keyboardsurfer commented on app/build.gradle.kts:
```diff
@@ -0,0 +1,108 @@
+
+        resConfig("en")
+        manifestPlaceholders = mapOf(
+                "crashlyticsEnabled" to false
```
Too long for a single line?

pfmaggi replied:
nope. shrinked to a single line.

keyboardsurfer commented on app/build.gradle.kts:
```diff
@@ -0,0 +1,108 @@
+    buildTypes {
+        getByName("release") {
+            // There's a Dex Splitter issue when enabling DataBinding & proguard in dynamic features
+            // The temporary workaround is to disable shrinking
```
Do we have a bug to link to here?

pfmaggi replied:
Yes, added note:
This issue is tracked in: https://issuetracker.google.com/120517460

keyboardsurfer commented on buildSrc/src/main/java/AddCommonRepositories.kt:
```diff
@@ -0,0 +1,90 @@
+        includeGroup("org.sonatype.oss")
+        includeGroup("org.xerial")
+        includeGroup("net.ltgt.gradle.incap")
+
```
Remove new line

pfmaggi replied:
Done

keyboardsurfer commented on buildSrc/src/main/java/Libs.kt:
```diff
@@ -0,0 +1,72 @@
+    val TEST_JUNIT = "junit:junit:"
+    val TEST_MOCKITO_CORE = "org.mockito:mockito-core:"
+    val TEST_COROUTINES = "org.jetbrains.kotlinx:kotlinx-coroutines-test:"
+    val TEST_ESPRESSO_CONTRIB = "androidx.test.espresso:espresso-contrib:"
```
AX_TEST...

pfmaggi replied:
TEST_AX_...

keyboardsurfer commented on buildSrc/src/main/java/Libs.kt:
```diff
@@ -0,0 +1,72 @@
+
+@Suppress("MayBeConstant") // Improve perf when changing values
+
+object Libs {
```
What's the sort order here?

pfmaggi replied:
It's now sorted with the constant names.

keyboardsurfer commented on depconstraints/build.gradle.kts:
```diff
@@ -0,0 +1,121 @@
+    id("maven-publish")
+}
+
+val appcompat = "1.1.0"
```
Why arent' these constant?

pfmaggi replied:
I would prefer to not put them in an object: "const val are only allowed on top level or in objects

keyboardsurfer commented on buildSrc/src/main/java/Libs.kt:
```diff
@@ -0,0 +1,72 @@
+@Suppress("MayBeConstant") // Improve perf when changing values
+
+object Libs {
+    val AX_ANNOTATION = "androidx.annotation:annotation:"
```
Either const & SCREAMING_SNAKE_CASE or non const & pascalCase

pfmaggi replied:
const & screaming

keyboardsurfer commented on gradle.properties:
```diff
@@ -16,17 +16,14 @@ org.gradle.parallel=true
+product_hunt_developer_token = "<your product hunt developer token>"
 android.useAndroidX=true
-android.enableJetifier=true
+android.enableJetifier=true
```
+\n

pfmaggi replied:
done

keyboardsurfer reviewed (changes requested):
AFAICT `depconstraints` only hosts a build file. Does it need to be its own project for that?

pfmaggi committed:
Address PR comments

pfmaggi reviewed (commented):
This is a special components that needs to be consumed by the other modules:
https://docs.gradle.org/current/userguide/java_platform_plugin.html#sec:java_platform_consumption

e.g. it needs to be a project that can be imported:
```
dependencies {
    // get recommended versions from the platform project
    api(platform(project(":platform")))
    // no version required
    api("commons-httpclient:commons-httpclient")
}
```

pfmaggi committed:
Address comments from the PR.
